### ISSUE TITLE
Add interface to Java API for reading/writing SimpleITK Image to/from JNI buffers

### ISSUE THREAD

**Issue by haesleinhuepf** [SimpleITK/SimpleITK#1215]
Hi all,

it appears that the interface for generating an `Image` from a JNI buffer (e.g. FloatBuffer) is not available.
Similarily, also image.getBufferAsFloat() is not exposed as well. These two interfaces would allow us to use SimpleITK from [ImageJ](https://imagej.github.io) and [CLIJ](https://clij.github.io) very conveniently.

To make a long story short, it would be cool if this would work in Java:
```
// make a JNI FloatBuffer with a small image
float[] float_array = new flot[]{1,2,3,4,5,6,7,8};
FloatBuffer in_buffer = FloatBuffer.wrap(float_array);
long[] dimensions = new long[]{2, 2, 2}; // 3D image 2x2x2 pixels

// import
Image image = SimpleITK.importImage(in_buffer, dimensions);

// export
FloatBuffer out_buffer = image.getBufferAsFloat();
```

See also:
https://discourse.itk.org/t/simpleitk-connecting-to-other-image-processing-libraries/3491

I'm not so much a C++-JNI expert, but I'm happy to help if possible.

Thanks a lot for your awesome work! I <3 SimpleITK

Cheers,
Robert

**Comment by blowekamp** [comment]
In PR #1227 another method `Image::GetBufferAsByteBuffer` is added, which returns a java.nio.ByteBuffer object. This use of the Java Buffer's instead of the "long" pointer representation needs to be evaluated to determine if it's just as efficient.

This packaging with the new change should be automatically completed tonight.

**Comment by haesleinhuepf** [comment]
@blowekamp 

I just tested the new GetBufferAsByteBuffer in simpleitk-2.1.0.dev54.jar and it works like a charm! I can also do toFloatBuffer on the resulting object and get the correct pixels out. So all good from my side!

I will update the corresponding ImageJ example code asap.

Thanks a lot for your support!

**Comment by blowekamp** [comment]
@haesleinhuepf Glad to hear it works smoothly and hopefully as efficiently! I think the Buffer classes and much better interface that the low level pointer and long integer presented. 

I am working on a[ PR](1238 ) to replace the other GetBufferAs methods to return different buffer types. There are some questions about the interface under the TODO section, that your input would be very much appreciated.

- Should the be renamed such as `GetBufferAsFloat32` to `GetBuferAsFloatBuffer` for
example?
- What to do about the C++ "const" methods. The java Buffer object
does have a "readonly" attribute, but Java doen't have a concept of
constant member functions. Additionally the constant methods don't call Image::MakeUnique, which can cause a buffer duplication when lazy coping has occurred.
-  Should buffer types not supported by Java should be removed?

Thanks for the input!

**Comment by haesleinhuepf** [comment]
@blowekamp Interesting questions. 

> * Should the be renamed such as `GetBufferAsFloat32` to `GetBuferAsFloatBuffer` for
>   example?

In Java they are called `FloatBuffer`. Thus, the term `image.getAsFloatBuffer` is reasonable for Java developers. `Float32Buffer` would be more precise.

> * What to do about the C++ "const" methods.

Not sure which methods that are...

> * Should buffer types not supported by Java should be removed?

I'm using multiple native-C++ libraries from Java. I can in most cases retrieve long-pointers from library A and send it to library B. The more open APIs from any library, the easier it is to connect them. Thus, it might make sense to keep stuff that's not directly compatible to Java. See also our similar discussion with the Java-OpenCL (JOCL) maintainer [here](https://github.com/gpu/JOCL/pull/26).

My suggestion: 
* Keep one consistent way of low-level API access. For example, provide a way to access long-pointers to images and furthermore methods to retrieve their pixel-type and dimensions. In that way one can use it with maximum flexibility.
* Furthermore, keep one high-level API access. For example, a method which returns a Java NIO Buffer. If this buffer is a FloatBuffer or a ByteBuffer or whatever depends on the image and is handled internally. The method just garantees the returned object is derived from [Java NIO Buffer](https://docs.oracle.com/javase/7/docs/api/java/nio/Buffer.html). In that way, maximum convenience is given.

Thanks for all your efforts. It's soo good to see the Java-ImageJ-SimpleITK-ITK bridge evolving.

**Comment by mkitti** [comment]
Following

**Comment by blowekamp** [comment]
This issue has been resolved with the merge of #1238.
